pkg/serviceaccount/tokens_controller.go

```diff
@@ -0,0 +1,384 @@
+			UpdateFunc: e.secretUpdated,
+			DeleteFunc: e.secretDeleted,
+		},
+	)
+
+	return e
+}
+
+// TokensController manages ServiceAccountToken secrets for ServiceAccount objects
+type TokensController struct {
+	client     *client.Client
+	token      TokenGeneratorFunc
+	kubeconfig KubeconfigGeneratorFunc
+
+	serviceAccountStore cache.Store
+	secretStore         cache.Store
+
+	// Since we join two objects, we'll watch both of them with controllers.
+	serviceAccountController *framework.Controller
+	secretController         *framework.Controller
+}
+
+// Runs controller loops and returns immediately
+func (e *TokensController) Run(stopCh <-chan struct{}) {
+	defer util.HandleCrash()
+	go e.serviceAccountController.Run(stopCh)
+	go e.secretController.Run(stopCh)
+}
+
+// serviceAccountAdded reacts to a ServiceAccount creation by creating a corresponding ServiceAccountToken Secret
+func (e *TokensController) serviceAccountAdded(obj interface{}) {
+	serviceAccount := obj.(*api.ServiceAccount)
+	e.createSecretIfNeeded(serviceAccount)
+}
+
+// serviceAccountUpdated reacts to a ServiceAccount update (or re-list) by ensuring a corresponding ServiceAccountToken Secret exists
+func (e *TokensController) serviceAccountUpdated(oldObj interface{}, newObj interface{}) {
+	newServiceAccount := newObj.(*api.ServiceAccount)
+	e.createSecretIfNeeded(newServiceAccount)
+}
+
+// serviceAccountDeleted reacts to a ServiceAccount deletion by deleting all corresponding ServiceAccountToken Secrets
+func (e *TokensController) serviceAccountDeleted(obj interface{}) {
+	serviceAccount, ok := obj.(*api.ServiceAccount)
+	if !ok {
+		// Unknown type. If we missed a ServiceAccount deletion, the
+		// corresponding secrets will be cleaned up during the Secret re-list
+		return
+	}
+	for _, secret := range e.listSecrets(serviceAccount) {
+		e.deleteSecret(secret)
+	}
+}
+
+// secretAdded reacts to a Secret creation by generating a API token (if the referenced user exists)
+// or by deleting the secret (if the referenced user does not exist)
+func (e *TokensController) secretAdded(obj interface{}) {
+	secret := obj.(*api.Secret)
+	if serviceAccount := e.getServiceAccount(secret); serviceAccount != nil {
+		e.createTokenIfNeeded(serviceAccount, secret)
+	} else if secret.Type == api.SecretTypeServiceAccountToken {
+		e.deleteSecret(secret)
+	}
+}
+
+// secretUpdated reacts to a Secret update (or re-list) by generating a API token (if the referenced user exists)
+// or by deleting the secret (if the referenced user does not exist)
+func (e *TokensController) secretUpdated(oldObj interface{}, newObj interface{}) {
+	oldSecret := oldObj.(*api.Secret)
+	newSecret := newObj.(*api.Secret)
+	if newServiceAccount := e.getServiceAccount(newSecret); newServiceAccount != nil {
+		e.createTokenIfNeeded(newServiceAccount, newSecret)
+	} else if newSecret.Type == api.SecretTypeServiceAccountToken {
+		e.deleteSecret(newSecret)
+	}
+
+	// Handle case where they changed the type or annotations
```

Comment:
You need resourceVersion as well as UID to reliably detect changes, since these objects are not immutable.